build-support: fix nix-prefetch-* on macOS

jdelStrother opened the pull request:
Since nix 2.20, `nix-store --add-fixed` doesn't accept paths where the parent directory is a symlink. On macOS, /tmp is a symlink to /private/tmp, which causes a `'/tmp' is a symlink` error:

```
$ nix run github:nixos/nixpkgs/24.11-beta#nix-prefetch-git -- --url https://github.com/IFTTT/polo.git --rev 316aa2ac210a45a7fc400ab921831493d5dd21b8 --hash sha256
Initialized empty Git repository in /private/tmp/git-checkout-tmp-1Bf9bIv7/polo-316aa2a/.git/
remote: Enumerating objects: 51, done.
remote: Counting objects: 100% (51/51), done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 51 (delta 8), reused 19 (delta 5), pack-reused 0 (from 0)
Unpacking objects: 100% (51/51), 19.57 KiB | 541.00 KiB/s, done.
From https://github.com/IFTTT/polo
 * branch            HEAD       -> FETCH_HEAD
Switched to a new branch 'fetchgit'
removing `.git'...
error: path '/tmp' is a symlink
```

Avoid this by resolving /tmp to a real directory in all the prefetch scripts

Fixes #358113

---

I don't see mention of the symlink-change in the nix 2.20 release notes, but I tested with this to narrow down the point where things started breaking:

```
nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/8b27c1239e5c421a2bbc2c65d52e4a6fbf2ff296.tar.gz \
  --packages 'nix-prefetch-git.override { nix = pkgs.nixVersions.nix_2_20; }' \
  --command 'nix-prefetch-git --url https://github.com/IFTTT/polo.git --rev 316aa2ac210a45a7fc400ab921831493d5dd21b8 --hash sha256'
```

Substitute nix_2_20 with nix_2_19 to see the old behavior which happily accepted symlinked directories.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jdelStrother committed:
build-support: fix nix-prefetch-* on macOS

Since nix 2.20, `nix-store --add-fixed` doesn't accept paths where the
parent directory is a symlink. On macOS, /tmp is a symlink to
/private/tmp, which causes a "'/tmp' is a symlink" error:

```
$ nix run github:nixos/nixpkgs/24.11-beta#nix-prefetch-git -- --url https://github.com/IFTTT/polo.git --rev 316aa2ac210a45a7fc400ab921831493d5dd21b8 --hash sha256
Initialized empty Git repository in /private/tmp/git-checkout-tmp-1Bf9bIv7/polo-316aa2a/.git/
remote: Enumerating objects: 51, done.
remote: Counting objects: 100% (51/51), done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 51 (delta 8), reused 19 (delta 5), pack-reused 0 (from 0)
Unpacking objects: 100% (51/51), 19.57 KiB | 541.00 KiB/s, done.
From https://github.com/IFTTT/polo
 * branch            HEAD       -> FETCH_HEAD
Switched to a new branch 'fetchgit'
removing `.git'...
error: path '/tmp' is a symlink
```

Avoid this by resolving /tmp to a real directory in all the prefetch scripts

emilazy commented on pkgs/build-support/docker/nix-prefetch-docker:
```diff
@@ -121,7 +121,7 @@ fi
 sourceUrl="docker://$imageName@$imageDigest"
 
-tmpPath="$(mktemp -d "${TMPDIR:-/tmp}/skopeo-copy-tmp-XXXXXXXX")"
+tmpPath="$(mktemp -d "$(realpath "${TMPDIR:-/tmp}/skopeo-copy-tmp-XXXXXXXX")")"
```
How about:
```suggestion
tmpPath="$(realpath "$(mktemp -d --tmpdir skopeo-copy-tmp-XXXXXXXX)")"
```
and analoguously for all the others?

jdelStrother replied:
This actually used to use --tmpdir over 10 years ago, but was changed to use TMPDIR in https://github.com/NixOS/nixpkgs/commit/5e7a1cf955e43e1fa3c1c157c0d6961e1bcf9801.

It does look like my /usr/bin/mktemp does support --tmpdir on macOS 15.1.1, but I'm not sure how far back that goes. I could add coreutils to all the prefetch scripts to try & make sure we've always got a usable `mktemp --tempdir` ..?

emilazy replied:
Ah, I knew I kept my trusty 10.12 VM around for something:

![Screenshot of macOS 10.12 confirming that its `mktemp` doesn’t support `--tmpdir`.](https://github.com/user-attachments/assets/237df9f4-aaed-4974-a7ad-064a2a472b32)

So this would be fine as‐is for 25.05 I think but not for 24.11. However, yes, we probably ought to just add `coreutils` to the list of common dependencies in `pkgs/tools/package-management/nix-prefetch-scripts/default.nix`; all our other Bash scripts assume GNU coreutils anyway.

emilazy reviewed (dismissed):
I think this is just a Nix bug. We can work around it, but I’d like to see a comment next to the occurrences pointing to the bug so we can remove it later when it’s fixed.

jdelStrother commented:
> I think this is just a Nix bug. We can work around it, but I’d like to see a comment next to the occurrences pointing to the bug so we can remove it later when it’s fixed.

OK, will try and get the nix devs to confirm whether it's a bug or feature so at least we have something concrete to link to.

jdelStrother committed:
build-support: Simplify tmpdir creation with coreutils

macOS 10.12 doesn't have a usable --tmpdir flag on the builtin mktemp, but we can
make use of coreutil's mktemp instead.

jdelStrother commented:
I think this is an intentional change rather than bug, judging by https://github.com/NixOS/nix/commit/83c067c0fa0cc5a2dca440e5c986afe40b163802. Shall I just update the comments to point to that commit ?

emilazy commented:
I think that rather that’s just an API change that they didn’t properly update all the callers of. I still think it is a Nix bug.

NobbZ commented:
I agree with emilazy. This breaks something that previously worked. And adding something that is behind a symlink (but not necessarily the symlink itself!) seems a valid usecase to me.

will commented:
I think this PR would fix what is currently breaking `crystal2nix`, but I haven't yet verified.

It starts doing its work, then gets `error: path '/tmp' is a symlink` when calling `nix-prefetch-git` , then fails to parse the empty stdout response as json

```
❯ crystal2nix
Initialized empty Git repository in /private/tmp/git-checkout-tmp-UCUG3iTb/backtracer.cr/.git/
From https://github.com/sija/backtracer.cr
 * tag               v1.2.2     -> FETCH_HEAD
Switched to a new branch 'fetchgit'
removing `.git'...
error: path '/tmp' is a symlink
Unhandled exception: Expected BeginObject but was EOF at line 1, column 1
  parsing Crystal2Nix::PrefetchJSON at line 0, column 0 (JSON::SerializableError)
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'raise<JSON::SerializableError>:NoReturn'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'Crystal2Nix::Worker#run:Nil'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in '__crystal_main'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'main'
Caused by: Expected BeginObject but was EOF at line 1, column 1 (JSON::ParseException)
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'raise<JSON::ParseException>:NoReturn'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'JSON::PullParser#raise<String>:NoReturn'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'Crystal2Nix::Worker#run:Nil'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in '__crystal_main'
  from /nix/store/fpnllrcmwbhpj2l3njin35jjb2f7jfdb-crystal2nix-0.3.0/bin/.crystal2nix-wrapped in 'main'
```

jdelStrother commented:
Yeah, sounds like the same issue.
I've re-written https://github.com/NixOS/nix/issues/11941 to try & clarify it's not just some macOS weirdness.

Mic92 commented:
Tested those prefetchers and they seem to work. Will bring up the bug in Nix again in the next meeting.

will commented:
Thanks!

Mic92 commented:
Work in progress pull request in nix that attempts to fix this: https://github.com/NixOS/nix/pull/12046
